text/0000-fabric-gateway.md

```diff
@@ -0,0 +1,213 @@
+
+Implementation options:
+- Only direct queries (and endorsement requests) to the peer(s) with the highest ledger height.
+- Client proposals include last observed ledger height so Gateway can target any peers with at least that ledger height.
+ A view of ledger height can be tracked by the Gateway from block events it has received.
+
+## Scaling and load balancing
+
+Currently a client application is responsible for load balancing its requests between multiple peers.  The Gateway will handle this on behalf of the set of currently connected client applications.
+
+Ideally, the Gateway itself will be stateless (i.e. not maintain client session state) allowing clients to be routed though an appriopriate load balancer.
+
+## Gateway SDKs
+
+A set of SDKs will be created to allow client applications to interact with the Fabric Gateway.  The APIs exposed by these SDKs will be, where possible, the same as the current high-level 'gateway' SDKs.
+
+The following classes/structures will be available as part of the gateway programming model:
+- Gateway
+- Network
+- Contract
+- Transaction
+- Wallet
+
+The existing `submitTransaction()` and `evaluateTransaction()` API methods will be implemented using the Gateway gRPC services described above.
+
+Based on user feedback of the existing gateway SDKs, the following additions are planned to be supported by this implementation:
+- Support for 'offline signing', allowing clients to write REST servers that don't directly have access to users' private keys.
+- Support for 'asynchronous submit', allowing client applications to continue processing while the transaction commit proceeds in the background. For Go this will be implemented by returning a channel; for Node by returning a Promise; and for Java by returning a Future.
+
+SDKs will be created for the following languages:
+- Node (Typescript/Javascript)
+- Go
+- Java
+
+It is proposed that these SDKs will be maintained in the same GitHub respository as the gateway itself (`hyperledger/fabric-gateway`) to ensure that they all stay up to date with each other and with the core gateway.  This will be enforced in the CI pipeline by running the end-to-end scenario tests across all SDKs.
+
+Publication of SDK releases will be as follows:
+- Node SDK published to NPM.
+- Java SDK published to Maven Central.
+- Go SDK published to separate GitHub repo (e.g. fabric-gateway-go) and tagged.
+
+Contributors are invited to create SDKs for other languages, e.g. Python, Rust.
+
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+When run as a standalone process, the gateway would add another docker containter to an (arguably) already complex Fabric architecture.  Also it will require its own signing identity in the channel writers policy in order to invoke the discovery service, which might be a security concern in some environments. These concerns are mitigated by embedding the gateway server within the organisation's existing peers.  
```

Comment:
Please remove the "docker container" reference. None of this is bound to docker.